ast/ast_test.go

```diff
@@ -0,0 +1,247 @@
+package ast_test
+
+import (
+	"context"
+	"io"
+	"testing"
+
+	"git.sr.ht/~nelsam/hel/v4/pkg/pers"
+	"github.com/earthly/earthly/ast"
+	"github.com/poy/onpar/v2"
+	"github.com/poy/onpar/v2/expect"
+)
+
+func TestParse(t *testing.T) {
+	type testCtx struct {
+		t      *testing.T
+		expect expect.Expectation
+		reader *mockNamedReader
+	}
+
+	o := onpar.BeforeEach(onpar.New(t), func(t *testing.T) testCtx {
+		return testCtx{
+			t:      t,
+			expect: expect.New(t),
+			reader: newMockNamedReader(t, timeout),
+		}
+	})
+
+	o.Spec("it safely ignores comments outside of documentation", func(tt testCtx) {
+		pers.ConsistentlyReturn(t, tt.reader.NameOutput, "Earthfile")
+		handleMockFile(tt.t, tt.reader, []byte(`
+# this is an early comment.
+
+# VERSION does not get documentation.
+VERSION 0.6 # Trailing comments do not cause parsing errors at the top level
+WORKDIR /tmp
+
+bar:
+
+baz:
+    # comments in an otherwise empty target should be
+    # ignored.
+
+# foo - Comments between targets should not be parsed as
+# documentation, even if they start with the target's name.
+
+foo: # inline  comments do not consume newlines
+    # RUN does not get documentation.
+    RUN echo foo
+
+    ARG foo=bar # inline comments should also be ignored.
+
+    # Lonely comment blocks in
+    # targets should be ignored.
+
+    # Even if they don't have a trailing newline.`))
+		f, err := ast.ParseOpts(context.Background(), ast.FromReader(tt.reader))
+		tt.expect(err).To(not(haveOccurred()))
+
+		tt.expect(f.Targets).To(haveLen(3))
+		foo := f.Targets[2]
+		tt.expect(foo.Name).To(equal("foo"))
+		tt.expect(foo.Docs).To(equal(""))
+	})
+
+	o.Spec("targets with leading whitespace cause errors", func(tt testCtx) {
+		pers.ConsistentlyReturn(t, tt.reader.NameOutput, "Earthfile")
+		handleMockFile(tt.t, tt.reader, []byte(`
+VERSION 0.6
+
+  foo:
+    RUN echo foo
+`))
+		_, err := ast.ParseOpts(context.Background(), ast.FromReader(tt.reader))
+		tt.expect(err).To(haveOccurred())
+		tt.expect(err.Error()).To(containSubstring("no viable alternative at input '  '"))
+	})
+
+	o.Spec("it parses a basic target", func(tt testCtx) {
+		pers.ConsistentlyReturn(t, tt.reader.NameOutput, "Earthfile")
+		handleMockFile(tt.t, tt.reader, []byte(`
+VERSION 0.6
+
+foo:
+    RUN echo foo
+`))
+		f, err := ast.ParseOpts(context.Background(), ast.FromReader(tt.reader))
+		tt.expect(err).To(not(haveOccurred()))
+
+		tt.expect(f.Version.Args).To(haveLen(1))
+		tt.expect(f.Version.Args[0]).To(equal("0.6"))
+
+		tt.expect(f.Targets).To(haveLen(1))
+		tgt := f.Targets[0]
+		tt.expect(tgt.Name).To(equal("foo"))
+		tt.expect(tgt.Recipe).To(haveLen(1))
+		rcp := tgt.Recipe[0]
+		tt.expect(rcp.Command).To(not(beNil()))
```

Comment:
ultra-nit:

This completely comes down to preferemce, but I can't say I'm a fan of this (rspec style?) testing library.

`To(not(beNil()))` doesn't feel very go-like (which could be a compliment depending on who you ask :laughing: )

We already use the `testify` library in the code base, which wouldn't require any nesting or chaining, e.g. `assert.NotNil(t, rcp.Command)`.

Is it worth standardising on a single testing framework? 

If an existing framework doesn't support what we need (or makes it cumbersome), then we should obviously re-evaluate it.

At the same time I don't want to get distracted on a tabs-vs-spaces discussion :smile:

Reply:
I was one of the very early contributors to testify, and I have a lot of problems with it. Some of my problems with it are simply problems I have with testing in go in general, but many are easily solved. One particularly easy one: in testify, it's very difficult to know which function parameter is the `expected` and which is the `actual`. I also recall some situations where the order of the parameters had to be reversed for some reason (probably something about variadic functions).

I started using rspec-style assertions in 2015 with [`ginkgo`](https://onsi.github.io/ginkgo/) and [`gomega`](https://onsi.github.io/gomega/). I was not a fan at first, but they've certainly grown on me. While I still think that `gomega` (and especially `ginkgo`) are over-complicated, I really appreciate that I never have a question about which parameter is the `actual` and which is the `expected`. And while `onpar/expect` isn't particularly heavily used, `gomega` _is_ - so I don't think the style is completely foreign in the go world any more.

I haven't found an assertion style that I _really_ like, but while I've grown more and more frustrated with testify-style assertion functions over time, I have definitely grown _less_ frustrated with rspec/gomega/expect style expectation matchers. `goconvey`-style `So(actual, ShouldEqual, expected)` doesn't look awful to me either, but I am dubious about `goconvey` because last I checked, it imports `gls` (goroutine-local-storage). I don't like having that in my import graph.

I still _really_ want to find a good assertion style that can make use of type parameters to avoid the "untyped nil" problem. I [tried to do it in `onpar`](https://github.com/poy/onpar/pull/35) but it made the built-in matchers feel awful.

---

For this PR, I'd be willing to go through and replace all the `tt.expect(...)` with `if ... { tt.t.Fatalf(...) }` instead.  I would _prefer_ to avoid using testify, personally, but if you _really_ like it I'll work through my dislike of that library.

Alternatively, we could wrap up `tt.expect(...).To(...)` in different verbiage. Maybe `tt.assert(...).Does(...)`?

Reply:
> One particularly easy one: in testify, it's very difficult to know which function parameter is the expected and which is the actual

I have definitely mixed them up before.

>  also recall some situations where the order of the parameters had to be reversed for some reason 

interesting point! I see `WithinRange` is an example where actual comes first (and then there's a range of expected vals).

Reply:
Some good arguments for both. Let's talk about this briefly in our Core sync-up to get to a resolution.